File: week_2/example4-pipe.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/types.h>
#include <sys/wait.h>


int main() {
    int pipefd[2];
    pid_t pid;

    if (pipe(pipefd) == -1) {
        perror("pipe");
        return 1;
    }

    pid = fork();

    if (pid < 0) {
        perror("fork");
        return 1;
    }

    if (pid == 0) {
        close(pipefd[1]);

        FILE *output = fopen("destination.txt", "w");

        char ch2;
        // read from pipe and convert to lowercase
        while (read(pipefd[0], &ch2, 1) > 0) {
            if (ch2 >= 'A' && ch2 <= 'Z') {
                ch2 = ch2 + 32;
            }
            fputc(ch2, output);
        }

        fclose(output);
        close(pipefd[0]);

    } else {
        close(pipefd[0]); 

        FILE *input = fopen("source.txt", "r");

        char ch1;
        while ((ch1 = fgetc(input)) != EOF) {
            // write to pipe
            write(pipefd[1], &ch1, 1);
        }

        fclose(input);
        close(pipefd[1]);

        wait(NULL);
    }

    return 0;
}

File: week_2/example4-pipe.py
#!/usr/bin/env python3
"""
Python equivalent of the C pipe/fork example
Demonstrates IPC using multiprocessing.Pipe()
"""

import multiprocessing as mp
import os

def child_process(pipe_end, output_file):
    """Child process: reads from pipe and converts uppercase to lowercase"""
    print(f"Child process PID: {os.getpid()}")

    with open(output_file, 'w') as dest:
        while True:
            # Read one character at a time from pipe
            data = pipe_end.recv()
            if data == 'EOF':  # End of file marker
                break

            ch = data
            # Convert uppercase to lowercase (same logic as C code)
            if 'A' <= ch <= 'Z':
                ch = chr(ord(ch) + 32)  # Add 32 to ASCII value

            dest.write(ch)

    pipe_end.close()
    print("Child process finished")

def parent_process(pipe_end, input_file):
    """Parent process: reads from file and writes to pipe"""
    print(f"Parent process PID: {os.getpid()}")

    with open(input_file, 'r') as source:
        while True:
            ch = source.read(1)  # Read one character
            if not ch:  # End of file
                break

            # Send character through pipe
            pipe_end.send(ch)

    # Send end-of-file marker
    pipe_end.send('EOF')
    pipe_end.close()
    print("Parent process finished")

def main():
    """Main function - equivalent to C main()"""
    input_file = "source.txt"
    output_file = "destination.txt"

    print("Python Pipe IPC Demo")
    print("===================")

    # Create pipe (equivalent to pipe() system call)
    parent_conn, child_conn = mp.Pipe()

    # Create child process (equivalent to fork())
    child = mp.Process(target=child_process, args=(child_conn, output_file))
    child.start()

    # Parent process work
    parent_process(parent_conn, input_file)

    # Wait for child to finish (equivalent to wait())
    child.join()

    print("Both processes completed!")
    print(f"Check '{output_file}' for the converted text")

if __name__ == "__main__":
    main()
